pkg/controller/gc/gc_controller.go

```diff
@@ -0,0 +1,140 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+const (
+	fullResyncPeriod = 0
+	gcCheckPeriod    = 30 * time.Second
+)
+
+type GCController struct {
+	kubeClient client.Interface
+	podControl controller.PodControlInterface
+	podStore   cache.StoreToPodLister
+	threshold  int
+}
+
+func New(kubeClient client.Interface, threshold int) *GCController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	gcc := &GCController{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "pod-garbage-collector"}),
+			KubeClient: kubeClient,
+		},
+		threshold: threshold,
+	}
+
+	terminalSelector, err := fields.ParseSelector(strings.Join([]string{
+		"status.phase!=" + string(api.PodPending),
+		"status.phase!=" + string(api.PodRunning),
+		"status.phase!=" + string(api.PodUnknown),
+	}, ","))
+
+	if err != nil {
+		panic("terminalSelector must compile")
+	}
+
+	gcc.podStore.Store, _ = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return gcc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), terminalSelector)
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return gcc.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), terminalSelector, rv)
+			},
+		},
+		&api.Pod{},
+		fullResyncPeriod,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	return gcc
+}
+
+func (gcc *GCController) Run(stop <-chan struct{}) {
+	util.Until(gcc.gc, time.Second, stop)
+}
+
+func (gcc *GCController) gc() {
+	terminalPods, _ := gcc.podStore.List(labels.Everything())
+	terminalPodCount := len(terminalPods)
+	sort.Sort(byCreationTimestamp(terminalPods))
+
+	deleteCount := terminalPodCount - gcc.threshold
+
+	if deleteCount > terminalPodCount {
+		deleteCount = terminalPodCount
+	}
+	if deleteCount > 0 {
+		glog.Infof("garbage collecting %v pods", deleteCount)
+	}
+
+	var wait sync.WaitGroup
+	for i := 0; i < deleteCount; i++ {
+		wait.Add(1)
+		go func(namespace string, name string) {
+			defer wait.Done()
+			if err := gcc.podControl.DeletePod(namespace, name); err != nil {
+				// ignore not founds
+				defer util.HandleError(err)
+			}
+		}(terminalPods[i].Namespace, terminalPods[i].Name)
+	}
+	wait.Wait()
+}
+
+func isPodTerminal(pod *api.Pod) bool {
```

Comment:
would also be great to define the string representing this in a centralized place